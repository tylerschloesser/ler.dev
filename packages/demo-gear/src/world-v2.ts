import { cloneDeep } from 'lodash-es'
import invariant from 'tiny-invariant'
import { initBeltPaths, initTiles } from './derived.js'
import {
  AddEntityError,
  Either,
  Vec2,
} from './types-common.js'
import { Derived } from './types-derived.js'
import { BuildEntity, Entity } from './types-entity.js'
import { Origin } from './types-origin.js'
import { World } from './types-world.js'

export function initWorld(): World {
  const origin = initOrigin()
  const derived = initDerived(origin)
  invariant(derived.right)
  return {
    origin,
    derived: derived.right,
    nextEntityId: 0,
  }
}

export function tryAddEntities(
  world: World,
  entities: BuildEntity[],
): Either<AddEntityError[], World> {
  if (entities.length === 0) {
    return { left: null, right: world }
  }

  const errors = validateEntitiesToAdd(entities)
  if (errors.length) {
    return { left: errors, right: null }
  }

  let { nextEntityId } = world
  const origin = cloneDeep(world.origin)

  for (const entity of entities) {
    addEntity(origin, {
      id: `${nextEntityId++}`,
      ...entity,
    })
  }

  const derived = initDerived(origin)
  if (derived.left) {
    return derived
  }
  invariant(derived.right)

  return {
    left: null,
    right: {
      origin,
      derived: derived.right,
      nextEntityId,
    },
  }
}

function getExistingEntity(
  origin: Origin,
  entity: BuildEntity,
): Entity | null {
  return null
}

function validateEntitiesToAdd(
  entities: BuildEntity[],
): AddEntityError[] {
  return []
}

function addEntity(origin: Origin, entity: Entity): void {
  invariant(!origin.entities[entity.id])
  origin.entities[entity.id] = entity
}

function initOrigin(): Origin {
  return {
    entities: {},
  }
}

function initDerived(
  origin: Origin,
): Either<AddEntityError[], Derived> {
  const tiles = initTiles(origin)
  const beltPaths = initBeltPaths(origin, tiles)
  if (beltPaths.left) {
    return beltPaths
  }
  invariant(beltPaths.right)
  return {
    left: null,
    right: {
      beltPaths: beltPaths.right,
      tiles,
    },
  }
}

function isVec2Equal(a: Vec2, b: Vec2): boolean {
  return a[0] === b[0] && a[1] === b[1]
}
