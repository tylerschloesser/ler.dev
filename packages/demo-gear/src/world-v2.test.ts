import { describe, expect, test } from '@jest/globals'
import { vec2 } from 'gl-matrix'
import invariant from 'tiny-invariant'
import {
  DerivedError,
  DerivedErrorType,
  Vec2,
} from './types-common.js'
import {
  Entity,
  entityType,
  BeltEntity,
} from './types-entity.js'
import { initWorld, tryAddEntities } from './world-v2.js'

function newBelt(
  args: Partial<BeltEntity> = {},
): BeltEntity {
  return {
    id: '0',
    type: entityType.enum.Belt,
    items: [],
    position: [0, 0],
    size: [1, 1],
    velocity: 0,
    offset: 0,
    ...args,
  }
}

function newBelts(positions: vec2[]): Entity[] {
  const entities: Entity[] = []
  for (let i = 0; i < positions.length; i++) {
    const position = positions.at(i)
    invariant(position !== undefined)
    entities.push(
      newBelt({
        id: `${i}`,
        // TODO why is this cast needed?
        position: position as Vec2,
      }),
    )
  }
  return entities
}

describe('world-v2', () => {
  describe('initWorld', () => {
    test('new world', () => {
      const actual = initWorld()
      expect(actual).toMatchSnapshot()
    })
  })

  describe('tryAddEntities', () => {
    test('empty', () => {
      const world = initWorld()
      const { origin, derived } = world
      const entities: Entity[] = []
      expect(tryAddEntities(world, entities)).toBeNull()

      expect(origin).toBe(world.origin)
      expect(derived).not.toBe(world.derived)

      expect(world).toMatchSnapshot()
    })

    test('add belt', () => {
      const world = initWorld()
      const entities: Entity[] = [newBelt()]
      expect(tryAddEntities(world, entities)).toBeNull()
      expect(world).toMatchSnapshot()
    })

    test('replace belt', () => {
      const world = initWorld()
      const entities: Entity[] = [newBelt()]
      expect(tryAddEntities(world, entities)).toBeNull()
      expect(tryAddEntities(world, entities)).toBeNull()
      expect(world).toMatchSnapshot()
    })

    test('add two disconnected belts', () => {
      const world = initWorld()
      const entities = newBelts([
        [0, 0],
        [2, 0],
      ])
      expect(tryAddEntities(world, entities)).toBeNull()
      expect(world).toMatchSnapshot()
    })

    test('add two connected horizontal belts', () => {
      const world = initWorld()
      const entities = newBelts([
        [0, 0],
        [1, 0],
      ])
      expect(tryAddEntities(world, entities)).toBeNull()
      expect(world).toMatchSnapshot()
    })

    test('more than two adjacent belts error', () => {
      const world = initWorld()
      const entities = newBelts([
        [0, 0],
        [1, 0],
        [2, 0],
        [1, 1],
      ])
      const expected: DerivedError = {
        type: DerivedErrorType.BeltHasMoreThanTwoAdjacentBelts,
        entityId: '1',
      }
      expect(tryAddEntities(world, entities)).toEqual(
        expected,
      )
      expect(world).toMatchSnapshot()
    })
  })
})
