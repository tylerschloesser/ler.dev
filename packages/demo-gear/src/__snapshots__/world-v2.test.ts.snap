// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`world-v2 initWorld new world 1`] = `
{
  "derived": {
    "beltPaths": [],
    "nextEntityId": 0,
    "tiles": {},
  },
  "origin": {
    "entities": {},
  },
}
`;

exports[`world-v2 tryAddEntities add belt 1`] = `
{
  "derived": {
    "beltPaths": [
      {
        "entities": [
          {
            "direction": "WestEast",
            "id": "0",
            "invert": false,
          },
        ],
        "loop": false,
      },
    ],
    "nextEntityId": 0,
    "tiles": {
      "0.0": {
        "entityIds": [
          "0",
        ],
      },
    },
  },
  "origin": {
    "entities": {
      "0": {
        "id": "0",
        "items": [],
        "offset": 0,
        "position": [
          0,
          0,
        ],
        "size": [
          1,
          1,
        ],
        "type": "Belt",
        "velocity": 0,
      },
    },
  },
}
`;

exports[`world-v2 tryAddEntities add two connected horizontal belts 1`] = `
{
  "derived": {
    "beltPaths": [
      {
        "entities": [
          {
            "direction": "WestEast",
            "id": "1",
            "invert": false,
          },
          {
            "direction": "WestEast",
            "id": "0",
            "invert": false,
          },
        ],
        "loop": false,
      },
    ],
    "nextEntityId": 0,
    "tiles": {
      "0.0": {
        "entityIds": [
          "0",
        ],
      },
      "1.0": {
        "entityIds": [
          "1",
        ],
      },
    },
  },
  "origin": {
    "entities": {
      "0": {
        "id": "0",
        "items": [],
        "offset": 0,
        "position": [
          0,
          0,
        ],
        "size": [
          1,
          1,
        ],
        "type": "Belt",
        "velocity": 0,
      },
      "1": {
        "id": "1",
        "items": [],
        "offset": 0,
        "position": [
          1,
          0,
        ],
        "size": [
          1,
          1,
        ],
        "type": "Belt",
        "velocity": 0,
      },
    },
  },
}
`;

exports[`world-v2 tryAddEntities add two connected vertical belts 1`] = `
{
  "derived": {
    "beltPaths": [
      {
        "entities": [
          {
            "direction": "WestEast",
            "id": "1",
            "invert": false,
          },
          {
            "direction": "WestEast",
            "id": "0",
            "invert": false,
          },
        ],
        "loop": false,
      },
    ],
    "nextEntityId": 0,
    "tiles": {
      "0.0": {
        "entityIds": [
          "0",
        ],
      },
      "0.1": {
        "entityIds": [
          "1",
        ],
      },
    },
  },
  "origin": {
    "entities": {
      "0": {
        "id": "0",
        "items": [],
        "offset": 0,
        "position": [
          0,
          0,
        ],
        "size": [
          1,
          1,
        ],
        "type": "Belt",
        "velocity": 0,
      },
      "1": {
        "id": "1",
        "items": [],
        "offset": 0,
        "position": [
          0,
          1,
        ],
        "size": [
          1,
          1,
        ],
        "type": "Belt",
        "velocity": 0,
      },
    },
  },
}
`;

exports[`world-v2 tryAddEntities add two disconnected belts 1`] = `
{
  "derived": {
    "beltPaths": [
      {
        "entities": [
          {
            "direction": "WestEast",
            "id": "0",
            "invert": false,
          },
        ],
        "loop": false,
      },
      {
        "entities": [
          {
            "direction": "WestEast",
            "id": "1",
            "invert": false,
          },
        ],
        "loop": false,
      },
    ],
    "nextEntityId": 0,
    "tiles": {
      "0.0": {
        "entityIds": [
          "0",
        ],
      },
      "2.0": {
        "entityIds": [
          "1",
        ],
      },
    },
  },
  "origin": {
    "entities": {
      "0": {
        "id": "0",
        "items": [],
        "offset": 0,
        "position": [
          0,
          0,
        ],
        "size": [
          1,
          1,
        ],
        "type": "Belt",
        "velocity": 0,
      },
      "1": {
        "id": "1",
        "items": [],
        "offset": 0,
        "position": [
          2,
          0,
        ],
        "size": [
          1,
          1,
        ],
        "type": "Belt",
        "velocity": 0,
      },
    },
  },
}
`;

exports[`world-v2 tryAddEntities empty 1`] = `
{
  "derived": {
    "beltPaths": [],
    "nextEntityId": 0,
    "tiles": {},
  },
  "origin": {
    "entities": {},
  },
}
`;

exports[`world-v2 tryAddEntities more than two adjacent belts error 1`] = `
{
  "derived": {
    "beltPaths": [],
    "nextEntityId": 0,
    "tiles": {},
  },
  "origin": {
    "entities": {
      "0": {
        "id": "0",
        "items": [],
        "offset": 0,
        "position": [
          0,
          0,
        ],
        "size": [
          1,
          1,
        ],
        "type": "Belt",
        "velocity": 0,
      },
      "1": {
        "id": "1",
        "items": [],
        "offset": 0,
        "position": [
          1,
          0,
        ],
        "size": [
          1,
          1,
        ],
        "type": "Belt",
        "velocity": 0,
      },
      "2": {
        "id": "2",
        "items": [],
        "offset": 0,
        "position": [
          2,
          0,
        ],
        "size": [
          1,
          1,
        ],
        "type": "Belt",
        "velocity": 0,
      },
      "3": {
        "id": "3",
        "items": [],
        "offset": 0,
        "position": [
          1,
          1,
        ],
        "size": [
          1,
          1,
        ],
        "type": "Belt",
        "velocity": 0,
      },
    },
  },
}
`;

exports[`world-v2 tryAddEntities replace belt 1`] = `
{
  "derived": {
    "beltPaths": [
      {
        "entities": [
          {
            "direction": "WestEast",
            "id": "0",
            "invert": false,
          },
        ],
        "loop": false,
      },
    ],
    "nextEntityId": 0,
    "tiles": {
      "0.0": {
        "entityIds": [
          "0",
        ],
      },
    },
  },
  "origin": {
    "entities": {
      "0": {
        "id": "0",
        "items": [],
        "offset": 0,
        "position": [
          0,
          0,
        ],
        "size": [
          1,
          1,
        ],
        "type": "Belt",
        "velocity": 0,
      },
    },
  },
}
`;
